### ISSUE TITLE
(Nio|Epoll)EventLoop.pendingTasks should not block when called outside of event loop

### ISSUE THREAD

**Issue by vkostyukov** [netty/netty#8196]
### Expected behavior

`EventLoop.pendingTasks` should be (reasonably) cheap to invoke so it can be used within observability.

### Actual behavior

We've been experimenting with overriding `EventExecutorChooserFactory` with more superior load balancing algorithms and found out that it might be quite non-trivial to support pending-tasks load metric (pick the least loaded event loop).

After poking around the code, it seems when querying `pendingTasks` of either `NioEventLoop`  or `EpollEventLoop`, the method [blocks if called outside of said event loop](https://github.com/netty/netty/blob/630c82717dd8d822323c9e4a8ad51ea4f8a38f86/transport/src/main/java/io/netty/channel/nio/NioEventLoop.java#L262-L272). Put it this way, it's only cheap to query event loop's pending tasks from within the same event loop. Perhaps I miss some (historical) context for `pendingTask`, but for an uneducated reader, it just defeats the purpose of a method providing visibility into an event loop.

I was hoping to pick up some context on the current behavior of `pendingTasks` and see if we ever want/will to change it.

**Comment by normanmaurer** [comment]
@vkostyukov thats a good point... We may want to "replace" the current implementation with an `AtomicInteger` that is updated . WDYT ?

**Comment by vkostyukov** [comment]
@normanmaurer Yeah, that seems reasonable to me. Even if we won't be able to guarantee certain prevention it should be totally fine for the sake of observability (provided we document it).

I'd be happy to contribute a fix for this unless somebody else want to work on this.

**Comment by normanmaurer** [comment]
@vkostyukov that would be awesome... Like you already stated we should just make this "as cheap as possible".

**Comment by normanmaurer** [comment]
Perfect... let her know I am always happy to help 👍

**Comment by Wingfril** [comment]
I'm not really sure why we even need to do that? Can't we just get the super.size() itself without worrying if it is in the event loop? It seems `MpscQueue` is no longer in the code, and so live locking shouldn't be an issue anymore? `SingleThreadEventExecutor` uses a `LinkedBlockingQueue` which has a size variable which is an `atomicInteger` already?

**Comment by johnou** [comment]
@Wingfril it's still in the code, see `io.netty.channel.nio.NioEventLoop#newTaskQueue`, `io.netty.channel.epoll.EpollEventLoop#newTaskQueue` and `io.netty.channel.kqueue.KQueueEventLoop#newTaskQueue`

**Comment by njhill** [comment]
@johnou actually I think @Wingfril is right, the original issue from which the run-in-eventloop workaround stemmed was based on a livelock issue (https://github.com/netty/netty/issues/3675, https://github.com/netty/netty/issues/5297) encountered with the old netty internal `MpscLinkedQueue` implementation used at the time

This was since replaced by the shaded jctools impl in https://github.com/netty/netty/pull/6295, and there's no reason to suspect its `size()` method suffers from the same problem.

So I would propose just to undo the original run-in-eventloop workaround and instead call `size()` on the queue directly, @normanmaurer do you agree?

**Comment by johnou** [comment]
@nitsanw is it safe to invoke MpscQueue.size() from multiple threads?

**Comment by nitsanw** [comment]
Yes

**Comment by normanmaurer** [comment]
Then please just remove the code 👍🔥

> Am 16.08.2018 um 16:27 schrieb Nitsan Wakart <notifications@github.com>:
> 
> Yes
>

**Comment by johnou** [comment]
@nitsanw was the yes for safe or risk producing a livelock? :)

**Comment by njhill** [comment]
@johnou I'm pretty sure the livelock issue wasn't inherently related to the single-consumerness. Looking at the history it appeared to be more likely linked to the custom modifications made (e.g. possibly `RecyclableMpscLinkedQueueNode`).

Of course the reported size may not be "precise" due to the lazy semantics but should hopefully be "safe".

**Comment by nitsanw** [comment]
You asked if it was safe, right?

**Comment by johnou** [comment]
@nitsanw and also asked if it was unsafe and posed risks ie. livelock, but given your comment now I will assume it's safe to invoke size() from multiple threads, thanks!

**Comment by nitsanw** [comment]
@johnou I apologize if I sounded impatient :-) I was reading your question from the GitHub email. Here's what I got:
> @nitsanw is it safe to invoke MpscQueue.size() from multiple threads?

It appears that further edits do not get sent as further e-mails.

So, yes, it's safe to call size on any queue in jctools from any thread, no issue. Furthermore, current queue being used is either the MpscChunkedArrayQueue or MpscUnboundedArrayQueue:
https://github.com/netty/netty/blob/06f3574e46e0623a45a93d36dbe7aa3d7455a995/common/src/main/java/io/netty/util/internal/PlatformDependent.java#L823

Both of which have a very fast `size()` method.

As a side note, I'm not sure why in the non-unsafe case you do not use MpscChunkedAtomicArrayQueue:
https://github.com/netty/netty/blob/06f3574e46e0623a45a93d36dbe7aa3d7455a995/common/src/main/java/io/netty/util/internal/PlatformDependent.java#L859

Thanks

**Comment by normanmaurer** [comment]
@nitsanw maybe consider doing a PR ? :)

**Comment by normanmaurer** [comment]
@vkostyukov fixes via https://github.com/netty/netty/pull/8197 by @Wingfril :)

**Comment by vkostyukov** [comment]
@normanmaurer We've been getting pretty mixed results so far (@Wingfril can probably speak more about it) as we weren't really able to employ the `pendingTask` method.

**Comment by normanmaurer** [comment]
@Wingfril I would very interested to hear more.

**Comment by Wingfril** [comment]
@normanmaurer 
Heyo. 
So we tried several different load metrics to try and apply power of 2 choices (and also a heap when using number of channels as a load metric), such as keeping track of the amount of channels in each eventloop, looking at the pending tasks per eventloop, and also I modified a local copy of netty to look at the awake time/alive time of the eventloop. 

In staging and in simulations, none of the metrics that we tried had improved overall latency or throughput. In fact, I don't think there was even a decrease in standard deviations between stats of the netty default and 2 of the load metrics. When we looked at the number of channels per eventloop, we were able to make the channels more evenly spread out to the eventloops, but again, this did not translate to performance improvements.
